docs: refactor get maas up and running (#209)
maas/maas: docs/how-to-guides/get-started/install-maas.md

# Install MAAS

> There are multiple approaches to installing MAAS, depending on your needs, use case, and operating system. This document outlines how to install MAAS on a single machine in region+rack mode. As MAAS is packaged without a database, both the MAAS snap and the PostgreSQL snap are required.

Reviewer: I would remove the part about PostgreSQL 

```suggestion
There are multiple approaches to installing MAAS, depending on your needs, use case, and operating system. This document outlines how to install MAAS on a single machine in region+rack mode.
```

> ## Requirements

Reviewer: We can list PostgreSQL as a requirement as well and mention that it can be installed via snap. What do you think?

Author: I think the later note to skip the section if already installed covers it, otherwise we'll have an optional requirement

A physical or virtual machine intended to run MAAS must meet the following requirements:

- Ubuntu 22.04 LTS (Jammy) or newer.
- `sudo` privileges.

```{admonition} Disable conflicting NTP services
:class: warning

In production environments, it is recommended to disable conflicting NTP services as MAAS will handle time synchronization of machines:
  ```bash
  sudo systemctl disable --now systemd-timesyncd
  ```

```{admonition} Older MAAS versions
:class: note

To install older versions of MAAS, see [Supported MAAS versions](/reference/supported-maas-versions.md) for compatible PostgreSQL channels and operating system versions. For this guide, we will use the latest version of MAAS.
```

## Install the MAAS snap

Install the MAAS snap with:

```bash

> sudo snap install maas --channel=3.7/stable

Reviewer: I am curious if we need to introduce `latest/stable` channel so the doc is always _up to date_
What do you think?

Author: This was on my list of points to discuss earlier actually: I think `latest/stable` would be great to have imo. But ultimately, we will always have to change the relevant documentation because of postgres version compatibility. 

We _could_ inject these in as variables in future, getting the MAAS version from `pyproject.toml`, but we'd need to add a more explicit variable for `PSQL_RECOMMENDED_VERSION`. We only have `PSQL_MIN_VERSION` at the moment in `plugin.py`, with 16 [hard-coded](https://github.com/canonical/maas/blob/8f4278e4e001daadf7e01446e20545279c08c6ad/src/maasserver/deprecations.py#L82) elsewhere

```

> ## Install PostgreSQL

Reviewer: I think this step is optional, someone might already have a PostgreSQL database up and running somewhere on the network.

Install the PostgreSQL snap:

```bash
sudo snap install postgresql --channel=16/stable
```

Configure credentials for the PostgreSQL database. Pick a strong password and store it securely:

```bash

> export DBPASS=<strong-password>

Reviewer: That already makes password insecure as it was exposed via ENV, I would rather propose a prompt approach here.

Author: Fixed, but this is much less "copy and paste" friendly. Other how-to instructions e.g. [aws](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html) are comfortable doing this though it's worth noting

Reviewer: Good point about copy-paste friendly. Oki, let's not bother too much and address this to TA

export DBUSER=maas
export DBNAME=maas
```

Create a MAAS database role and database:

```bash
sudo postgresql.psql -U postgres -c "CREATE USER \"$DBUSER\" WITH ENCRYPTED PASSWORD '$DBPASS'"

> sudo postgresql.createdb -U postgres -O "$DBUSER" "$DBNAME"

Reviewer: ```suggestion
postgresql.createuser -U postgres -h /tmp -P maas
postgresql.createdb -U postgres -h /tmp -O maas maas
```

```

## Initialize MAAS

> Initialize MAAS in region+rack mode with the PostgreSQL database credentials:

Reviewer: ```suggestion
Initialize MAAS in region+rack mode:
```

```bash

> sudo maas init region+rack --database-uri "postgres://$DBUSER:$DBPASS@localhost/$DBNAME"

Reviewer: ```suggestion
sudo maas init region+rack --database-uri "postgres://<user>:<password>@<hostname>/<dbname>"
```

Author: I have set user and dbname hardcoded to maas instead to align with your above suggestion, this way the user only needs to input one thing

```

Press Enter when prompted to accept the default MAAS URL, or specify your own.

## Create an administrator user

Create an administrator user, specifying the username, email and [SSH import id](https://manpages.ubuntu.com/manpages/noble/man1/ssh-import-id.1.html) to import a public key (optional):

```bash
sudo maas createadmin \
  --username admin \
  --email admin@localhost \
  --ssh-import gh:<github-username>  # optional
```

Specify the password for your new administrator user when prompted. Store it securely. 

> <!--- do we need them to store this securely or can it be obtained later? --->

Reviewer: We don't store password in clean text anywhere. Password can be reset if you have privileged access to `maas` CLI.

```
root@maas:~# maas changepassword --help
usage: maas changepassword [-h] [--database {default}] [username]

Change a MAAS user's password.

positional arguments:
  username              Username to change password for; by default, it's the current username.

options:
  -h, --help            show this help message and exit
  --database {default}  Specifies the database to use. Default is "default".
```

## Verify your installation

Verify the installation by creating a profile and logging in:

```bash

> maas login admin "http://localhost:5240/MAAS/api/2.0/" $(sudo maas apikey --username=admin)

Reviewer: ```suggestion
maas login admin "http://localhost:5240/MAAS" $(sudo maas apikey --username=admin)
```

I think that should be enough, no need for `/api/2.0/`

```

You should see the following message:

```{terminal}

ubuntu@host-system:~$ maas login admin "http://localhost:5240/MAAS/api/2.0/" $(sudo maas apikey --username=admin)

You are now logged in to the MAAS server at
http://localhost:5240/MAAS/api/2.0/ with the profile name 'admin'.

For help with the available commands, try:

  maas admin --help

```

```{admonition} Profiles
:class: note

> You have just created a profile in the MAAS CLI called `admin`, logged in using the credentials created in the previous step for your admin user. The rest of these docs will refer to this profile as `$PROFILE`.

Reviewer: ```suggestion
You have just created a profile in the MAAS CLI called `admin`, logged in using the API key created in the previous step for your admin user. The rest of these docs will refer to this profile as `$PROFILE`.
```

```

## Access the MAAS UI

> Open a browser and navigate to `http://<maas-ip>:5240/MAAS`, replacing `<maas-ip>` with the IP address of your MAAS machine. If you are on the same machine as the machine running MAAS, use `http://localhost:5240/MAAS`.

Reviewer: ```suggestion
Open a browser and navigate to `http://<maas-ip>:5240/MAAS`, replacing `<maas-ip>` with the IP address of your MAAS region controller. If you are on the same machine as the machine running MAAS, use `http://localhost:5240/MAAS`.
```
